experimental/sql/src/ibis_frontend.py

```diff
@@ -0,0 +1,104 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from modulefinder import Module
+from unittest import result
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+
+import ibis
+import ibis.expr.types
+import ibis.expr.datatypes
+import ibis.expr.operations.relations as rels
+import ibis.expr.operations.generic as gen_types
+from ibis.expr.operations.logical import Equals as EQ
+import ibis.backends.pandas.client as PandasBackend
+
+import dialects.ibis_dialect as id
+
+# This file contains the translation from ibis to the ibis_dialect.
+# It is built in a visitor-style, as this is one of the best designs
+# to crawl trees in my experience. The entry-point is the `visit`-
+# function just below. It dispatches to the appropriate translator.
+
+
+@dataclass
+class NodeVisitor:
```

Comment:
I rewrote everything using this package. It feels way nicer to me, particularly since the top-level visit, that just calls the others, is not needed anymore. Plus, I like shifting as much complexity to others as possible.

Reply:
Cool, yes, that looks better.

Question: is it such that *every* expression has *exactly one* `op`? Would it make sense to visit any expression type with the same visitor, take the op out, and then visit that op (which dispatches to op-specific implementations)?